src/databricks/labs/ucx/hive_metastore/grants.py

```diff
@@ -307,3 +327,208 @@ def grants(
+                if len(eligible_locations) == 0:
+                    continue
+                grant = self._get_grants(eligible_locations, principals, tables, mounts)
+                grants.extend(grant)
+        catalog_grants = [Grant(principal, "USE", "hive_metastore") for principal in principals]
+        grants.extend(catalog_grants)
+
+        return list(set(grants))
+
+    def _get_aws_grants(self) -> list[Grant]:
+        # TODO
+        return []
+
+    def _get_mount_location(self, location: str, mounts: list[Mount]):
+        for mount in mounts:
+            if location.startswith(f"dbfs:{mount.name}") or location.startswith(f"/dbfs{mount.name}"):
+                return mount.source
+        return None
+
+    def _get_privilege(self, table: Table, locations: dict[str, str], mounts: list[Mount]):
+        if table.view_text is not None:
+            # return nothing for view so that it goes to the separate view logic
+            return None
+        if table.location is None:
+            return None
+        if table.location.startswith('dbfs:/mnt') or table.location.startswith('/dbfs/mnt'):
+            mount_location = self._get_mount_location(table.location, mounts)
+            if mount_location is None:
+                return None
+            for loc, privilege in locations.items():
+                if loc is not None and mount_location.startswith(loc):
+                    return privilege
+            return None
+        if table.location.startswith('dbfs:/') or table.location.startswith('/dbfs/'):
+            return "WRITE_FILES"
+
+        for loc, privilege in locations.items():
+            if loc is not None and table.location.startswith(loc):
+                return privilege
+        return None
+
+    def _get_database_grants(self, tables: list[Table], principals: list[str]) -> list[Grant]:
+        databases = []
+        for table in tables:
+            if table.database not in databases:
+                databases.append(table.database)
+        return [
+            Grant(principal, "USE", "hive_metastore", database) for database in databases for principal in principals
+        ]
+
+    def _get_grants(
+        self, locations: dict[str, str], principals: list[str], tables: list[Table], mounts: list[Mount]
+    ) -> list[Grant]:
+        grants = []
+        filtered_tables = []
+        for table in tables:
+            privilege = self._get_privilege(table, locations, mounts)
+            if privilege == "READ_FILES":
+                grants.extend(
+                    [Grant(principal, "SELECT", table.catalog, table.database, table.name) for principal in principals]
+                )
+                filtered_tables.append(table)
+            if privilege == "WRITE_FILES":
+                grants.extend(
+                    [
+                        Grant(principal, "ALL PRIVILEGES", table.catalog, table.database, table.name)
+                        for principal in principals
```

Comment:
This is better as a dedicated method on ExternalLocations class (or the external location migration class) - eg given a Table instance, return an ExternalLocation.

Reply:
Seperated to AzureACL. will do similar way for AWS as separate PR